Update to latest PromiseKit

jakubknejzlik opened the pull request:
Hello guys 👋

There's bunch of code changes, but the tests finally passed! Thank you for your review in advance. Change requests expected :)

tldr:
- PromiseKit is completely refactored out as extension
- methods now return `CodableStoreRequest` which is going to be used for cancellation 
- any idead how to write `CodableStoreRequest.send` method nicer are welcome 🙏

Commit:
Update to latest PromiseKit, resolve #13

VilemKurz commented on CodableStore/Classes/Environment.swift:
```diff
@@ -8,6 +8,7 @@
 import Foundation
 import PromiseKit
 
+
```
space

tomassliz reviewed (changes requested):
I've discovered two issues in this update:

1. UI API is called on a background thread: `UIApplication.setNetworkActivityIndicatorVisible`
2. Deserialization of objects doesn't set correctly `dateDecodingStrategy` & `keyDecodingStrategy`. That's because casting on line 58 & 59 in `Serialization.swift` always fails:
```swift
public func deserialize<T: Decodable>() throws -> T {
    if let data = self as? T {
        return data
    } else {
        let decoder = JSONDecoder()

        // THIS TYPE CASTING ALWAYS FAILS
        decoder.dateDecodingStrategy = (T.self as? CustomDateDecodable.Type)?.dateDecodingStrategy ?? .iso8601
        decoder.keyDecodingStrategy = (T.self as? CustomKeyDecodable.Type)?.keyDecodingStrategy ?? .useDefaultKeys

        return try decoder.decode(T.self, from: self)
    }
}
```
But tests don't catch this issue so the should be updated too.

Commit:
Fix: UI API called from background thread

Summary:
- Synchronize write and read using DispatchQueue
- Async UI to main queue

Commit:
Fix failing tests

Summary: Not sure why, but API was giving me constantly different ids
so I refused to fight it and changed the ids of requests. :)

Commit:
Update to Swift 4.2

Summary:
- Update example project
- Update pods

Commit:
Tests serialize and deserialize

Summary:
- Enable code coverage
- Add test that verifies if custom strategy is used during the (de)serialization

jakubpetrik commented:
@tomassliz I've added a test to see if custom strategy is used or not and it indeed is. Can you please give some minimal example in which it is failing for you? Also network indicator should be now fixed.

Commit:
Remove extra space

Summary: Remove empty line from Environmnet.swift

3ducation1 commented:
@jakubpetrik I will try to post some example during this weekend.

tomassliz commented:
@jakubpetrik I will try to post some example during this weekend.

tomassliz commented:
@jakubpetrik I'm a little bit stuck. I can't replicate errors in tests but I have project which is failing. I can send you the project source with more info on email. Are you ok with that?

jakubpetrik commented:
> @jakubpetrik I'm a little bit stuck. I can't replicate errors in tests but I have project which is failing. I can send you the project source with more info on email. Are you ok with that?

@tomassliz  sure. :)

tomassliz commented:
> @tomassliz sure. :)

@jakubpetrik OK, can you send me your e-mail please? Just write me on iCloud.com email joined with my GitHub username.

jakubpetrik commented:
Ok. I can now confirm that issue with getting custom strategies still exists. 
Issue is when `T` is `Optional<CustomDateDecodable>`. I will try to fix that.

Commit:
Fix: Serialization casting always fails

Summary:
- Add FakeAPIEnvironment in test target
- Update serialization tests to catch the issue
- Move models from Environment tests to separate files
- Fix issue in CodableStoreRequest where deserialize type was infered as Optional
- Add Serialization conformances to Array, Dictionary and Optional

tomassliz commented:
@jakubpetrik @jakubknejzlik I verified changes and it's ok now 👍  Shall we merge this PR?

jakubpetrik commented:
Let's merge it please. @jakubknejzlik
